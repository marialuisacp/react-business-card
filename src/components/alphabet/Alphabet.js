import React, { Component } from 'react';
import { copyObject, getNextLetter } from '../../utils';
import { updateLetter } from '../../actions';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import './Alphabet.scss';
import '../../styles/styles.scss';

class Alphabet extends Component {
  constructor(props) {
    super(props);
    this.state = {
      slides: [
        { order: -1, id: 1, text: "" },
        { order: 0, id: 2, text: "D" },
        { order: 1, id: 3, text: "E" },
        { order: 2, id: 4, text: "G" },
        { order: 3, id: 5, text: "J" },
        { order: 4, id: 6, text: "L" }
      ],
      slideCount: 3,
      sliding: false,
      direction: 0,
      visibleCenterLetter: 'D',
      initPlaying: false
    };
  }

  handleNext = () => {
    let slidesCopy = {};
    if (!this.state.initPlaying) {
      slidesCopy = copyObject(this.state.slides);
      slidesCopy[0].text = 'P';
    }
    this.setState((prevState) => {
      return {
        sliding: true,
        direction: 1,
        visibleCenterLetter: getNextLetter(prevState.visibleCenterLetter, 1),
        initPlaying: true,
        slides: Object.keys(slidesCopy).length ? slidesCopy : prevState.slides
      }
    },
      () => {
        setTimeout(() => {
          this.setState((prevState) => {
            let oldSlides = [...prevState.slides];
            let hiddenSlide = oldSlides.shift();
            let newHiddenSlide = {
              ...hiddenSlide,
              order: oldSlides[oldSlides.length - 1].order + 1
            };
            oldSlides.push(newHiddenSlide);
            const slides = oldSlides.map(s => {
              return { ...s, order: s.order - 1 };
            });
            return { slides, sliding: false };
          });
        }, 500);
      });
  };

  handlePrev = () => {
    this.setState((prevState) => {
      return {
        sliding: true,
        direction: -1,
        visibleCenterLetter: getNextLetter(prevState.visibleCenterLetter, -1),
      }
    }, () => {
      setTimeout(() => {
        this.setState((prevState) => {
          let slides = [...prevState.slides];
          let lastSlide = slides.pop();
          lastSlide = { ...lastSlide, order: -1 };
          slides = slides.map(s => {
            return { ...s, order: s.order + 1 };
          });
          slides.unshift(lastSlide);
          return { slides, sliding: false };
        });
      }, 500);
    });
  };

  componentDidUpdate(prevProps) {
    const { letterCenter: letter, direction: lastDirection } = this.props;
    if (prevProps.letterCenter !== letter) {
      console.log('trocou');
      if (lastDirection === 1) {
        this.handleNext();
      } else if (lastDirection === -1) {
        this.handlePrev();
      }
    }
  };

  render() {
    const { slides, initPlaying, sliding, direction } = this.state;
    const { letterCenter: letter, updateLetter } = this.props;

    const slideActionStyle = sliding
      ? (direction > 0
        && {
          transform: "translateX(-130px)",
          transition: "transform 500ms ease-in"
        })
      ||
      {
        transform: "translateX(130px)",
        transition: "transform 500ms ease-in"
      }
      : {};

    return (
      <div id="alphabet" className='content'>
        <div className='list-letters'>
          <div
            className="item-carousel"
            style={slideActionStyle}
          >
            {slides && slides.length && slides.map(({ order, id, text }, idx) => {
              return <div key={idx} className="item">{text}</div>;
            })}
          </div>
          {(initPlaying && <div className="left-btn btn" onClick={() => updateLetter(getNextLetter(letter, -1), -1)}>
          </div>) || ''}
          <div className="right-btn btn" onClick={() => updateLetter(getNextLetter(letter, 1), 1)}>
          </div>
        </div>
      </div>
    );
  }
};

const mapDispatchToProps = dispatch =>
  bindActionCreators({ updateLetter }, dispatch);

const mapStateToProps = store => ({
  letterCenter: store.letterState.currentLetter,
  direction: store.letterState.direction,
});

export default connect(mapStateToProps, mapDispatchToProps)(Alphabet);