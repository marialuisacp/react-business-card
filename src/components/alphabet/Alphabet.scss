
.container {
  height: 130px;
  width: 600px; /* meant to hold 3 items */
  margin: 20px auto;
  overflow: hidden;
}

.left-btn, .right-btn {
  position: absolute;
  width: 50px;
  height: 50px;
  color: limeGreen;
}

.left-btn i, .right-btn i {
 cursor: pointer; 
}

.left-btn {
  top: 100px;
  left: 50px;
}

.right-btn {
  top: 100px;
  right: 50px;
}

.item-carousel {
  position: relative;
  right: 0;
  font-size: 0;
  // margin-top: -150px;
  width: 200%;
  transform: translateX(-130px);
}

.item {
  box-sizing: border-box;
  display: inline-block;
  height: 130px;
  width: 130px;
  background-color: transparent;
  border: none;
  font-size: 36px;
  text-align: center;
  vertical-align: middle;
  line-height: 164px;
  user-select: none;
  
  &.active {
    border: 3px solid black;
  }
}

@media (max-width: 850px) {
  .container {
    width: 130px; /* meant to hold 2 items */
  }
}

.slide-enter {
  transform: translateX(0px);
}

list-letters {
  display: flex;
  width: 390px;

  margin: 0 auto;
  position: relative;

  .letter {
    display: inline-block;
    border: none;
    background: transparent;
    width: 130px;
    height: 130px;
    outline: none;
    text-align: center;
    margin: 0 auto;
  }
  .letter span {
    color: #fff;
    font-size: 32px;
    font-weight: 300;
  }
}