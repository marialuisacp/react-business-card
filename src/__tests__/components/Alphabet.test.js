import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Alphabet from '../../components/Alphabet/Alphabet';
import { Provider } from 'react-redux'

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Alphabet', () => {
  it('renders without crashing given the required props', () => {
    const props = {
      letterCenter: 'D',
      dispatch: jest.fn(),
      direction: 1
    }

    const store = mockStore({
      currentLetter: 'D',
      direction: 0
    })

    const wrapper = shallow(<Provider store={store}>
      <Alphabet {...props} />
    </Provider>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})