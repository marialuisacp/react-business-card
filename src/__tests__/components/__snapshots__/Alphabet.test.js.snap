// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alphabet renders without crashing given the required props 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(Alphabet)
    direction={1}
    dispatch={[MockFunction]}
    letterCenter="D"
  />
</ContextProvider>
`;
